modified:   data/products.js

LukshaMaria opened the pull request:
modified:   main.js
modified:   modules/charts.js
modified:   modules/maintable.js
modified:   modules/producttable.js
modified:   modules/userlist.js

LukshaMaria committed:
modified:   data/products.js
modified:   main.js
modified:   modules/charts.js
modified:   modules/maintable.js
modified:   modules/producttable.js
modified:   modules/userlist.js

LukshaMaria committed:
modified:   data/products.js
modified:   main.js
modified:   modules/maintable.js
modified:   modules/producttable.js
modified:   task-webix.html

LukshaMaria committed:
new file:   data/categories.js
new file:   data/countries.js
modified:   modules/maintable.js

LukshaMaria committed:
modified:   modules/maintable.js
changed Math.random()*100 to Math.random()*(max - min) + min;

LukshaMaria committed:
modified:   main.js
modified:   modules/maintable.js

LukshaMaria committed:
Merge branch 'main' into webix-task-6

HelgaListopad commented on main.js:
```diff
@@ -42,5 +42,34 @@ webix.ready(function () {
+        this.group({
+            by: "country",
+            map: {
+                age: ["age", "sum"]
```
> let it show the number of users per country. 

Тут нужен count, а не sum

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -42,5 +42,34 @@ webix.ready(function () {
+            }
+        });
+    });
+    $$("datatable1").sync($$(collection));
```
Нужно убрать 

а) Таблица не должна синхронизироваться с этой коллекцией. Синхронизация предназначена для того, чтобы передавать основной набор данных.
Таблица "datatable1" уже получает данные по `url: "./data/data.js"`, а в `collection`  не те данные.

б) К коллекции не нужно обращаться через `$$`.
Метод нужен для того, чтобы получать UI-элементы, а коллекция у тебя уже создана и доступна через имя переменной (`collection`)

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -42,5 +42,34 @@ webix.ready(function () {
+        {
+            columnId: "year", compare: function (value, filter, item) {
+                let year = value;
+                if (filter == 1) return year;
+                else if (filter == 2) return (year < 1980);
+                else if (filter == 3) return (year > 1980 && year < 2000);
+                else if (filter == 4) return year > 2000;
```
Для такого случая лучше подойдёт конструкция `switch`

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -42,5 +42,34 @@ webix.ready(function () {
+        },
+        {
+            getValue: function (node) {
+                return node.getValue();
```
По смыслу лучше переименовать `node` на `view` 
По логу можно проверить, что в данном случае это tabbar
`registerFilter` позволяет регистрировать как фильтр и HTML-элементы (например, input), и в таком случае было бы `node`, как в документации

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -42,5 +42,34 @@ webix.ready(function () {
+            getValue: function (node) {
+                return node.getValue();
+            },
+            setValue: function (node, value) {
```
node => view

LukshaMaria replied:
fixed

HelgaListopad commented on modules/charts.js:
```diff
@@ -1,4 +1,5 @@
 const userchart = {
+    id: "userchart",
     view: "chart",
     url: "./data/users.js",
```
Чарт синхронизирован с листом, поэтому своего собственного источника данных у него не должно быть (это приичина, по которой в чарте отображается слишком много столбиков)

LukshaMaria replied:
fixed

HelgaListopad commented on modules/maintable.js:
```diff
@@ -1,34 +1,65 @@
+                { id: "rating", header: "Rating", width: 100 },
+                { id: "votes", header: ["Votes", { content: "numberFilter" }], width: 100, sort: "int" },
+                { id: "rank", header: "Rank", width: 50 },
+                { id: "category", header: "Category", width: 100 },
```
Здесь нужно колонке добавить `collection`
```js
{ id: "category", header: "Category", width: 100, collection: collection },

// ES6 позволяет писать вот так, если имя параметра и переменной совпадают
{ id: "category", header: "Category", width: 100, collection },
```

LukshaMaria replied:
fixed

HelgaListopad commented on modules/maintable.js:
```diff
@@ -1,34 +1,65 @@
+            ],
+            scheme: {
+                $init:function(obj){
+                    const max=5;
+                    const min =1;
+                    obj.category=Math.floor(Math.random()*(max - min) + min);                    
```
Можно сократить и сразу подставить числа, не объявляя лишние переменные
По code style - не стоит забывать про отступы и пробелы

LukshaMaria replied:
fixed

HelgaListopad commented on modules/producttable.js:
```diff
@@ -1,18 +1,32 @@
+    editor: "text",
+    editValue: "title",
     url: "./data/products.js",
+    rules: {
+        value: webix.rules.isNotEmpty,
+        value: (value) => (value.isNotNumber),
+    },
```
а) `rules` - это объект, а в объекте не может быть двух одинаковых ключей
б) есть ли в данных `value` и зачем его валидировать, если редактируется `title` ?

LukshaMaria replied:
title

LukshaMaria replied:
fixed

HelgaListopad commented on modules/producttable.js:
```diff
@@ -1,18 +1,32 @@
+    defaults: {
+            template: "{common.icon()} {common.folder()} #title#", width: 200
+     }
+    // columns: [
+    //     { id: "id", header: "", width: 50 },
+    //     {
+    //         id: "value", header: "Title",
+    //         template: "{common.icon()} {common.folder()} <span>#value#</span>", width: 200
+    //     },
+    //     { id: "price", header: "Price", width: 100 }
+    // ],
```
а) дерево - это не TreeTable, в нём не будет columns
б) ненужные комментарии из кода стоит убирать

LukshaMaria replied:
fixed

HelgaListopad commented on main.js:
```diff
@@ -42,5 +42,34 @@ webix.ready(function () {
+            map: {
+                age: ["age", "sum"]
+            }
+        });
```
Синхронизация срабатывает и на сортировку/фильтрацию листа, и при сортировке данные в чарте будут "скакать": группировка срабатывает по тем же критериям, но данные собираются в другом порядке
Чтобы избежать этого эффекта, после группировки к чарту можно применить свою сортировку по одному из имеющихся критериев (age, country)

HelgaListopad reviewed (changes requested):
В modules\forms.js измени текстовое поле для категорий
```js
{ view: "text", name: "category", label: "Category" },
```
На richselect c опциями из соответствующей коллекции (коллекция - то, что можно и нужно переиспользовать без webix.copy в нескольких виджетах)

LukshaMaria committed:
modified:   main.js
modified:   modules/charts.js
modified:   modules/forms.js
modified:   modules/maintable.js
modified:   modules/producttable.js

LukshaMaria committed:
Merge branch 'webix-task-6' of https://github.com/LukshaMaria/Projects-Webix into webix-task-6

LukshaMaria committed:
modified:   main.js
modified:   modules/maintable.js
modified:   modules/producttable.js

LukshaMaria committed:
modified:   main.js
modified:   modules/maintable.js

HelgaListopad commented on main.js:
```diff
@@ -46,28 +46,29 @@ webix.ready(function () {
-                else if (filter == 3) return (year > 1980 && year < 2000);
-                else if (filter == 4) return year > 2000;
+                switch (filter) {
+                    case 1: return year;
```
Фильтрация должна возвращать только boolean

LukshaMaria replied:
fixed, added value "true"

LukshaMaria committed:
modified:   main.js

LukshaMaria committed:
modified:   main.js

LukshaMaria committed:
modified:   main.js
new file:   modules/constvalues.js
modified:   modules/forms.js
modified:   modules/producttable.js
modified:   modules/userlist.js

LukshaMaria committed:
modified:   main.js
	modified:   modules/charts.js
	modified:   modules/forms.js
	modified:   modules/maintable.js
	modified:   modules/menu.js
	modified:   modules/producttable.js
	modified:   modules/toolbar.js
	modified:   modules/userlist.js
new file:   modules/values.js

LukshaMaria committed:
modified:   main.js
modified:   modules/menu.js
modified:   modules/userlist.js

LukshaMaria committed:
deleted:    modules/constvalues.js

LukshaMaria committed:
modified:   main.js
modified:   modules/forms.js
modified:   modules/maintable.js
modified:   modules/producttable.js
modified:   modules/toolbar.js
modified:   modules/userlist.js
modified:   modules/values.js
